### PR TITLE
Introduce result-clause

### LINKED ISSUE #40531: [Task]: Change Parser not to break the code when handling the ending clause of query expressions
### Description

When we introduce `collect-clause` as the ending clause of the query expression, there should be an API to retrieve it, and we have ton introduce it without breaking other places

### Describe your task(s)

_No response_

### Related area

-> Compilation

### Related issue(s) (optional)

_No response_

### Suggested label(s) (optional)

_No response_

### Suggested assignee(s) (optional)

_No response_

### PR DESCRIPTION
**Pull request by KavinduZoysa**
## Purpose

Fixes #40531

### CODE REVIEW

File: compiler/ballerina-parser/src/main/java/io/ballerina/compiler/syntax/tree/QueryExpressionNode.java [hunk 1]
```diff
@@ -43,11 +43,15 @@ public QueryPipelineNode queryPipeline() {
...
     @Deprecated
     public SelectClauseNode selectClause() {
-        ClauseNode endClause = endClause();
-        return endClause.kind() == SyntaxKind.SELECT_CLAUSE ? (SelectClauseNode) endClause : null;
+        ClauseNode endClause = resultClause();
+        if (endClause.kind() == SyntaxKind.SELECT_CLAUSE) {
+            return (SelectClauseNode) endClause;
+        } else {
+            throw new IllegalStateException("select-clause is not found");
+        }
```
Comment (rdulmina) [inline compiler/ballerina-parser/src/main/java/io/ballerina/compiler/syntax/tree/QueryExpressionNode.java:51]:
```suggestion
        if (endClause.kind() != SyntaxKind.SELECT_CLAUSE) {
            throw new IllegalStateException("select-clause is not found");
        } 
        return (SelectClauseNode) endClause;
```

**Comment by rdulmina** [comment]
We need to also change the names in `BallerinaParser.java` and `BallerinaErrorhandler.java` eg: https://github.com/ballerina-platform/ballerina-lang/blob/e1ef9867d68a0cadf149687bd24098254d6230e6/compiler/ballerina-parser/src/main/java/io/ballerina/compiler/internal/parser/BallerinaParserErrorHandler.java#L447

File: compiler/ballerina-parser/src/main/java/io/ballerina/compiler/syntax/tree/QueryExpressionNode.java [hunk 2]
```diff
@@ -43,11 +43,15 @@ public QueryPipelineNode queryPipeline() {
 
     @Deprecated
     public SelectClauseNode selectClause() {
-        ClauseNode endClause = endClause();
-        return endClause.kind() == SyntaxKind.SELECT_CLAUSE ? (SelectClauseNode) endClause : null;
+        ClauseNode endClause = resultClause();
+        if (endClause.kind() == SyntaxKind.SELECT_CLAUSE) {
+            return (SelectClauseNode) endClause;
+        } else {
+            throw new IllegalStateException("select-clause is not found");
```
[the remark this answers is not in the document]
Comment (KavinduZoysa) [reply compiler/ballerina-parser/src/main/java/io/ballerina/compiler/syntax/tree/QueryExpressionNode.java]: Changed

File: compiler/ballerina-parser/src/test/java/io/ballerinalang/compiler/parser/test/tree/DeprecatedSyntaxAPITest.java [hunk 3]
```diff
@@ -42,7 +42,11 @@ public void testQueryExprDeprecatedSelectClauseAPI() {
         Assert.assertNotNull(queryExpressionNode.selectClause());
 
         queryExpressionNode = (QueryExpressionNode) modulePart.findToken(113).parent().parent();
-        Assert.assertNull(queryExpressionNode.selectClause());
+        try {
+            queryExpressionNode.selectClause();
+        } catch (IllegalStateException e) {
```
[the remark this answers is not in the document]
Comment (KavinduZoysa) [reply compiler/ballerina-parser/src/test/java/io/ballerinalang/compiler/parser/test/tree/DeprecatedSyntaxAPITest.java:47]: Changed

File: compiler/ballerina-parser/src/main/java/io/ballerina/compiler/syntax/tree/QueryExpressionNode.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (rdulmina) [reply compiler/ballerina-parser/src/main/java/io/ballerina/compiler/syntax/tree/QueryExpressionNode.java]: Okay lets keep the existing msg

**Comment by rdulmina** [comment]
> > We need to also change the names in `BallerinaParser.java` and `BallerinaErrorhandler.java` eg:
> > https://github.com/ballerina-platform/ballerina-lang/blob/e1ef9867d68a0cadf149687bd24098254d6230e6/compiler/ballerina-parser/src/main/java/io/ballerina/compiler/internal/parser/BallerinaParserErrorHandler.java#L447
> 
> Changed

Did u check the `BallerinaParser.java` as well for endClause occurrences?

File: compiler/ballerina-parser/src/main/java/io/ballerina/compiler/syntax/tree/QueryExpressionNode.java [hunk 4]
```diff
@@ -43,11 +43,14 @@ public QueryPipelineNode queryPipeline() {
 
     @Deprecated
```
Comment (lochana-chathura) [inline compiler/ballerina-parser/src/main/java/io/ballerina/compiler/syntax/tree/QueryExpressionNode.java:44]: Not related to this PR. Small suggestion.
```suggestion
    /**
     * @deprecated This method is deprecated. Use {@link #resultClause()} instead.
     */
    @Deprecated
```

File: compiler/ballerina-parser/src/main/java/io/ballerina/compiler/syntax/tree/QueryExpressionNode.java [hunk 5]
```diff
@@ -43,11 +43,14 @@ public QueryPipelineNode queryPipeline() {
...
     public SelectClauseNode selectClause() {
-        ClauseNode endClause = endClause();
-        return endClause.kind() == SyntaxKind.SELECT_CLAUSE ? (SelectClauseNode) endClause : null;
+        ClauseNode endClause = resultClause();
+        if (endClause.kind() != SyntaxKind.SELECT_CLAUSE) {
+            throw new IllegalStateException("Query expression having collect-clause does not contain select-clause");
```
Comment (lochana-chathura) [inline compiler/ballerina-parser/src/main/java/io/ballerina/compiler/syntax/tree/QueryExpressionNode.java:48]: As per the concluded discussion: https://github.com/ballerina-platform/ballerina-lang/pull/40537#discussion_r1209971133
```suggestion
            throw new IllegalStateException("select-clause not found");
```

**Comment by lochana-chathura** [comment]
Need to rename the following variable.
https://github.com/ballerina-platform/ballerina-lang/blob/e1ef9867d68a0cadf149687bd24098254d6230e6/misc/formatter/modules/formatter-core/src/main/java/org/ballerinalang/formatter/core/FormattingTreeModifier.java#L2831

**Comment by lochana-chathura** [comment]
> Did u check the `BallerinaParser.java` as well for endClause occurrences?

Seems no occurrences there.

File: compiler/ballerina-parser/src/main/java/io/ballerina/compiler/syntax/tree/QueryExpressionNode.java (hunk 4 above)
Comment (KavinduZoysa) [reply compiler/ballerina-parser/src/main/java/io/ballerina/compiler/syntax/tree/QueryExpressionNode.java:47]: Added

File: compiler/ballerina-parser/src/main/java/io/ballerina/compiler/syntax/tree/QueryExpressionNode.java [hunk 6]
```diff
@@ -41,13 +41,19 @@ public QueryPipelineNode queryPipeline() {
...
     }
 
+    /**
+     * @deprecated Use {@link #resultClause()} instead.
+     */
     @Deprecated
     public SelectClauseNode selectClause() {
-        ClauseNode endClause = endClause();
-        return endClause.kind() == SyntaxKind.SELECT_CLAUSE ? (SelectClauseNode) endClause : null;
+        ClauseNode endClause = resultClause();
```
Comment (ushirask) [inline compiler/ballerina-parser/src/main/java/io/ballerina/compiler/syntax/tree/QueryExpressionNode.java:49]: Shouldn't `endClause` here be renamed as well?

Comment (KavinduZoysa) [reply compiler/ballerina-parser/src/main/java/io/ballerina/compiler/syntax/tree/QueryExpressionNode.java]: Yes, changed

**Review by lochana-chathura — APPROVED** [review]

**Pull request merged**